rllib/algorithms/algorithm.py

```diff
@@ -2303,8 +2240,112 @@ def _kwargs_for_execution_plan(self):
+
+        Args:
+            train_future: In case, we are training and avaluating in parallel,
+                this arg carries the currently running ThreadPoolExecutor
+                object that runs the training iteration
+
+        Returns:
+            The results dict from the evaluation call.
+        """
+        eval_results = {"evaluation": {}}
+        try:
+            if self.config["evaluation_duration"] == "auto":
+                assert (
+                    train_future is not None
+                    and self.config["evaluation_parallel_to_training"]
+                )
+                unit = self.config["evaluation_duration_unit"]
+                eval_results = self.evaluate(
+                    duration_fn=functools.partial(
+                        self._automatic_evaluation_duration_fn,
+                        unit,
+                        self.config["evaluation_num_workers"],
+                        self.config["evaluation_config"],
+                        train_future,
+                    )
+                )
+            # Run `self.evaluate()` only once per training iteration.
+            else:
+                eval_results = self.evaluate()
+
+        # In case of any failures, try to ignore/recover the failed evaluation workers.
+        except Exception as e:
+            self.try_recover_from_step_attempt(
+                error=e,
+                worker_set=self.evaluation_workers,
+                ignore=self.config["evaluation_config"].get("ignore_worker_failures"),
+                recreate=self.config["evaluation_config"].get(
+                    "recreate_failed_workers"
+                ),
+            )
+        return eval_results
+
+    def _run_one_training_iteration_and_evaluation_in_parallel(
+        self,
+    ) -> Tuple[ResultDict, "TrainIterCtx"]:
+        """Runs one training iteration and one evaluation step in parallel.
+
+        First starts the training iteration (via `self._run_one_training_iteration()`)
+        within a ThreadPoolExecutor, then runs the evaluation step in parallel.
+        In auto-duration mode (config.evaluation_duration=auto), makes sure the
+        evaluation step takes roughly the same time as the training iteration.
+
+        Returns:
+            The accumulated training and evaluation results.
+        """
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            train_future = executor.submit(lambda: self._run_one_training_iteration())
+            # Pass the train_future into `self._run_one_evaluation()` to allow it
+            # to run exactly as long as the training iteration takes in case
+            # evaluation_duration=auto.
+            results = self._run_one_evaluation(train_future)
+            # Collect the training results from the future.
+            train_results, train_iter_ctx = train_future.result()
+            results.update(train_results)
```

Comment:
Should have put my comment form above here: How to write a metric that includes both, training and evaluation results?